### PR TITLE
add autodiff expressions for common operations

### PR DESCRIPTION
**Pull request by zou3519**
This PR does a few things:

## Enable testing autodiff in test_jit.py

Previously test_jit.py only tested autograd on backward graphs.
This is because we borrow from test_autograd and construct graphs with a small
number of nodes. Because the number of nodes is small (typically 1-2), those graph
do not end up containing autodiff subgraphs, so autodiff never gets tested.

This PR enables autodiff testing by doing the following:
- added disableDebugAutodiffSubgraphInlining fn to graph_executor to disable
  autodiff subgraph inlining.
- (implementation) added autodiffSubgraphNodeThreshold and autodiffSubgraphInlineThreshold.
  These are set to their default values (2, 5) but disableDebugAutodiffSubgraphInlining()
  sets both to 1, disabling subgraph inlining and allowing 1-node autodiff subgraphs.
- The relevant backward jit tests disable autodiff subgraph inlining so they
  will test the autodiff versions of the operators instead of autograd whenever
  an autodiff variant exists.
- We don't run the tests that do inline autodiff subgraphs anymore.
  This has no impact on testing correctness because the assumption is
  that autograd functions are correct and are tested in test_autograd.py

## Add autodiff for a lot of ops

This allows the graph fuser to work better because a lot of these ops were previously not autodiff-compatible but fusible. On a more concrete example, lstm backward contains a lot of tensor-scalar operations; these autodiff formulas help its double backward pass.

Included:
- arithmetic overloads
- abs, acos, asin, atan, ceil, cos, cosh, exp, expm1, floor, fmod, frac, log, log10, log1p, log2 reciprocal, remainder, round, sin, sinh, tan, trunc, rsqrt

## Test Plan
TestJitGenerated tests autodiff for all of the added operations. 

cc @apaszke @zdevito

### CODE REVIEW

**Comment by zou3519** [comment]
There is something really fishing going on with LSTM's double backward (that is the test that is failing). Disabling the autodiff formula for mul(Tensor, Scalar) causes the test to pass, renabling it causes the test to fail. As far as I can tell the autodiff formula is correct so I'm not too sure what is going on (it's possible that it's a numeric precision issue, but the gradients are off by ~100 (they are on the order of magnitude of 1000)

**Comment by zou3519** [comment]
@apaszke thank you for pointing that out! (I would have gone on a long search for the cause otherwise...).

**Comment by zou3519** [comment]
Should be ready for review now. I excluded some advanced indexing tests from running because they fail autodiff tests and I don't see an easy fix for them. The root cause is #12024.

File: test/test_jit.py [hunk 1]
```diff
@@ -7732,12 +7745,24 @@ def new_fn(*tensors_):
     return new_fn, [arg for arg in args if isinstance(arg, torch.Tensor)]
 
 
+def assert_fully_differentiable_graph(graph):
+    nodes = list(graph.nodes())
+    if not (len(nodes) == 1 and 'prim::DifferentiableGraph' in str(nodes[0])):
```
Comment (apaszke) [inline test/test_jit.py]: can we check `nodes[0].kind() == 'prim::DifferentiableGraph` instead? Feels less hacky

File: torch/csrc/jit/graph_executor.cpp [hunk 2]
```diff
@@ -361,9 +361,20 @@ struct GraphExecutorImpl {
...
+    autodiffSubgraphNodeThreshold = 1;
+    // Don't inline autodiff subgraphs into autograd functions
+    autodiffSubgraphInlineThreshold = 1;
+  }
+
 private:
   friend struct GraphExecutor;
 
+  size_t autodiffSubgraphNodeThreshold = 2;
+  size_t autodiffSubgraphInlineThreshold = 5;
```
Comment (apaszke) [inline torch/csrc/jit/graph_executor.cpp]: Can you please move this to the bottom of this struct, where all other members are?

**Review by apaszke — APPROVED** [review]
Nice! Haven't reviewed all AD formulas, but they should be tested anyway

File: test/test_jit.py (hunk 1 above)
Comment (zou3519) [reply test/test_jit.py]: Oh I don't think I'm actually using this function. I'll delete if if I'm not (otherwise, your suggestion sounds good)

**Pull request closed without merging**